File: app/schemas/stock_in.py
"""Stock-in (Kirim) schemas."""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.product import ProductBrief


class StockInItemCreate(BaseModel):
    """One inbound line to add."""

    product_id: int
    quantity: Decimal = Field(gt=0)
    price: Decimal = Field(ge=0, description="Sotib olish narxi (birlik uchun)")


class StockInCreate(BaseModel):
    """Payload to create an inbound delivery document.

    ``store_id`` is required for a CEO (the receiving store) and omitted by a
    Seller (resolved from the token).
    """

    store_id: int | None = None
    supplier_id: int | None = None
    date: datetime | None = None
    note: str | None = None
    items: list[StockInItemCreate] = Field(min_length=1)


class StockInItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    quantity: Decimal
    price: Decimal
    subtotal: Decimal
    product: ProductBrief | None = None


class SupplierBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class UserBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    full_name: str


class StockInOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    reference: str
    store_id: int | None = None
    supplier_id: int | None = None
    created_by_id: int
    date: datetime
    total_amount: Decimal
    note: str | None = None
    created_at: datetime
    supplier: SupplierBrief | None = None
    created_by: UserBrief | None = None
    items: list[StockInItemOut] = Field(default_factory=list)

File: frontend/src/types/stock-in.ts
import type { ProductBrief, UserBrief } from "./common";

export interface StockInItem {
  id: number;
  product_id: number;
  quantity: string;
  price: string;
  subtotal: string;
  product: ProductBrief | null;
}

export interface StockIn {
  id: number;
  reference: string;
  store_id: number | null;
  created_by_id: number;
  date: string;
  total_amount: string;
  note: string | null;
  created_by: UserBrief | null;
  items: StockInItem[];
}

export interface StockInListParams {
  store_id?: number;
  search?: string;
  date_from?: string;
  date_to?: string;
}

export interface StockInItemFormValues {
  product_id: string;
  quantity: number;
  price: number;
}

export interface StockInFormValues {
  store_id?: string;
  date?: string;
  note?: string;
  items: StockInItemFormValues[];
}
